### ISSUE TITLE
add rule to enforce `FROM`, `JOIN` and `ON` statements to be on a separate line

### ISSUE THREAD

**Issue by jessie-carta** [sqlfluff/sqlfluff#1247]
Currently none of the rules enforce any behavior with `FROM`, `JOIN` and `ON` statements. For example, the below are currently acceptable:
```
FROM a
JOIN b ON a.id = b.id
```
```
FROM a JOIN
 b ON a.id = b.id
```

It would be nice to have a rule that enforces `FROM` and `JOIN` to be on new lines, and `ON` to be indented on a new line:
```
FROM a
JOIN b
  ON a.id = b.id
```

Similarly, this rule (or a separate rule) could enforce `AND`s in a join to be indented on a new line:
```
FROM a
JOIN b
  ON a.id = b.id
  AND a.id2 = b.id2
```
instead of 
```
FROM a
JOIN b
  ON a.id = b.id AND a.id2 = b.id2
```

**Comment by puetsch** [comment]
+1

I noticed the same when running it against the example below. Row 5 should be flagged as error IMO.

```sql
select
    dogs.*,
    cats.x,
    dogs.yo as yolo,
    count(distinct cats.x) as cnt from cats -- not OK
left join dogs on cats.id = dogs.id
group by 2, 3
having cnt > 2```

**Comment by danny-dailypay** [comment]
I'd like to piggyback on this and kindly request that it would be great if `GROUP BY`, `ORDER BY`, `PARTITION BY` also received similar optional functionality

**Comment by elyobo** [comment]
As discussed briefly over in https://github.com/sqlfluff/sqlfluff/issues/2600#issuecomment-1034760045, I'd also like something like this, but possibly not quite the same. Copying content from that discussion here...

<hr>
Sorry, to be clearer, the linter permits things like this

```sql
SELECT field,
  other_field,
```

but I would like to be able to configure it to _require_ that it be on a new line if there's more than one line in the block e.g.

```sql
-- fine because one line
SELECT field, other_field

-- fine because multiple lines but they're consistently indented under the SELECT
SELECT
  field,
  other_field,
```
<hr>

I don't really want to require another line all the time, but if a block _does_ have more than one line then everything except the block opener (e.g. WITH, SELECT (or SELECT DISTINCT), FROM, etc) must be on another line.

**Comment by eetu-h** [comment]
I created #2627 but it was closed as a duplicate of this one. What I'd like to see is to enforce rules for `where`, `having` and `qualify` clauses as well (similar to `select`) so that conditions should be on a new line, unless there is only one condition.

best practice 1:
 
```
select
    attr1,
    attr2
from db.tbl1
where
    attr3 = 'A' and
    attr4 = 'B'
```
    
best practice 2:
 
```
select
    attr1,
    attr2
from db.tbl1
where attr3 = 'A'
```

anti-pattern 1:
 
```
select
    attr1,
    attr2
from db.tbl1
where attr3 = 'A' and
    attr4 = 'B'
```
 
anti-pattern 2:
 
```
select
    attr1,
    attr2
from db.tbl1
where
    attr3 = 'A'
```
Would also be great to be able to configure consistent formatting for `from` and `joins` in addition to the previous. For example:
```
select
    tbl1.attr1,
    tbl1.attr2,
    tbl2.attr0
from
    db.tbl1
    inner join db.tbl2 on
        tbl1.attr1 = tbl2.attr1 and
        tbl1.attr2 = tbl2.attr2
where
    tbl1.attr3 = 'A' and
    tbl1.attr4 = 'B'
```

**Comment by eetu-h** [comment]
> Agree with most of that @eetu-h except the placement of those ANDs (which AFAIK isn't really at issue in this issue anyway?).

The placement of operators are defined in the [Rule_L007](https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.core.rules.Rule_L007) so the example below (or similar) should work as well:

``` sql
select
    tbl1.attr1,
    tbl1.attr2,
    tbl2.attr0
from
    db.tbl1
    inner join db.tbl2 on -- <on> could also be on the next line?
        tbl1.attr1 = tbl2.attr1
        and tbl1.attr2 = tbl2.attr2
where
    tbl1.attr3 = 'A' 
    and tbl1.attr4 = 'B'
```

**Comment by avijit3660** [comment]
Hi Was there a workaround implemented for this enhancement? Or we just did not got to this yet?

**Comment by yannik207** [comment]
Hey, I would also be interested in this feature :)

**Comment by Jaxsss** [comment]
Hi, are you looking to integrate this functionality?

**Comment by Garrettjson** [comment]
@yannik207 @Jaxsss I think you may be able to use some combination of these operators: `indented_on_contents`, `indented_using_on`, `allow_implicit_indents`.

From the [documentation](https://docs.sqlfluff.com/en/stable/configuration/layout.html#configuring-indent-locations)
